src/mlpack/tests/rectangle_tree_test.cpp

```diff
@@ -854,6 +864,244 @@ BOOST_AUTO_TEST_CASE(DiscreteHilbertValueTest)
                                                                   point4), -1);
 }
 
+template<typename TreeType>
+void CheckOverlap(const TreeType& tree)
+{
+  bool success = true;
+
+  // Check if two nodes overlap each other.
+  for (size_t i = 0; i < tree.NumChildren(); i++)
+  {
+    success = true;
+
+    for (size_t j = 0; j < tree.NumChildren(); j++)
+    {
+      if (j == i)
+        continue;
+      success = false;
+      // Two nodes overlap each other if and only if there are no dimension
+      // in which they do not overlap each other.
+      for (size_t k = 0; k < tree.Bound().Dim(); k++)
+      {
+        if ((tree.Child(i).Bound()[k].Lo() >=
+            tree.Child(j).Bound()[k].Hi()) ||
+            (tree.Child(j).Bound()[k].Lo() >=
+            tree.Child(i).Bound()[k].Hi()))
+        {
+          success = true;
+          break;
```

Comment:
Hm, I would expect `Intersect()` to actually return the bound intersection (an `HRectBound` not a `bool`) and I would expect `Overlap()` to return the amount of volume of the overlap.  I see your concern though, because it is not clear if `Contains()` means "contains entirely" or "contains partially".  We could also implement `Overlap()` to return the volume of the overlap and check if it is 0, too, and maybe that would work.

To the second point, I agree, we could add that functionality to `HRectBound` if it is duplicated.